django/db/models/fields/json.py

```diff
@@ -0,0 +1,377 @@
+            key_transforms.insert(0, previous.key_name)
+            previous = previous.lhs
+
+        lhs, params = compiler.compile(previous)
+        return lhs, params, key_transforms
+
+    def as_postgresql(self, compiler, connection):
+        lhs, params, key_transforms = self._preprocess_lhs(compiler, connection)
+        if len(key_transforms) > 1:
+            return "(%s %s %%s)" % (lhs, self.postgres_nested_operator), [key_transforms] + params
+        try:
+            int(self.key_name)
+        except ValueError:
+            lookup = "'%s'" % self.key_name
+        else:
+            lookup = "%s" % self.key_name
+        return "(%s %s %s)" % (lhs, self.postgres_operator, lookup), params
+
+    def as_mysql(self, compiler, connection):
+        lhs, params, key_transforms = self._preprocess_lhs(compiler, connection)
+        json_path = self.mysql_compile_json_path(key_transforms)
+        return 'JSON_EXTRACT(%s, %%s)' % lhs, params + [json_path]
+
+    def mysql_compile_json_path(self, key_transforms):
+        path = ['$']
+        for key_transform in key_transforms:
+            try:
+                num = int(key_transform)
+                path.append('[{}]'.format(num))
+            except ValueError:  # non-integer
+                path.append('.')
+                path.append(key_transform)
+        return ''.join(path)
+
+
+class KeyTextTransform(KeyTransform):
+    postgres_operator = '->>'
+    postgres_nested_operator = '#>>'
+
+
+class KeyTransformTextLookupMixin:
+    """
+    Mixin for combining with a lookup expecting a text lhs from a JSONField
+    key lookup. Make use of the ->> operator instead of casting key values to
+    text and performing the lookup on the resulting representation.
+    """
+    def __init__(self, key_transform, *args, **kwargs):
+        assert isinstance(key_transform, KeyTransform)
+        if builtin_connection.vendor == 'postgresql':
+            KeyTextTransform.output_field = TextField()
+
+        key_text_transform = KeyTextTransform(
+            key_transform.key_name, *key_transform.source_expressions, **key_transform.extra
+        )
+        super().__init__(key_text_transform, *args, **kwargs)
+
+    def process_lhs(self, compiler, connection):
+        lhs, lhs_params = super().process_lhs(compiler, connection)
+        if connection.vendor == 'mysql':
+            return 'JSON_UNQUOTE(%s)' % lhs, lhs_params
+        return lhs, lhs_params
+
+
+class CaseInsensitiveMixin:
+    def process_lhs(self, compiler, connection):
+        lhs, lhs_params = super().process_lhs(compiler, connection)
+        if connection.vendor == 'mysql':
+            return 'LOWER(%s)' % lhs, lhs_params
+        return lhs, lhs_params
+
+    def process_rhs(self, compiler, connection):
+        rhs, rhs_params = super().process_rhs(compiler, connection)
+        if connection.vendor == 'mysql':
+            return 'LOWER(%s)' % rhs, rhs_params
+        return rhs, rhs_params
```

Comment:
Do we need this mixin? It should be already handled by `PatternLookup` that uses `connection.pattern_ops`.

Reply:
For some reason `LIKE` and `LIKE BINARY` doesn't work very well with JSON values on MySQL.

Reply:
Or maybe it was the regex functions that had issues with case-sensitivity and JSON values, I forgot. I'll recheck later.

Reply:
Yep, `LIKE` doesn't work as expected. See the last two queries: https://dbfiddle.uk/?rdbms=mariadb_10.2&fiddle=4396a0f03cace7eb13d8e3bfdf83944f